---
title: "Assessment 4_SLE777_R Project"
author: "Ameeta"
date: "2025-10-05"
output: pdf_document
---

Gene expression 
Step 1: Download file for geneexpression.tsv

The link for raw file were copied from the github and then downloaded in R-markdown using te download.file function and then the file was downloaded locally with the name geneexpression.tsv using destfile argument. 
The file was sucessfully downloaded. 
```{r}
# download the gene expression file

download.file("https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/gene_expression.tsv", destfile =  "geneexpression.tsv") # destfile = destination file path 
```

Step 1: Reading gene expression file with gene identifiers as the row names and dispaying first 6 genes 

The downloaded file was read using the read.table function, where the header argument specifies that the first row contain the name of the columns, while sep= "\t" indicates that the file is in TSV format. The row.names =1 was used to set the first column as the names of gene identifiers, stringAsFactors=FALSE ensures that the data are represented as plain texts rather than factors. The obtained dataset was saved in the object gene_data. Following that, to inspect the contents, the head function was used display the first 6 rows, showing 6 gene identifiers.The value 6 was used to ensure onky 6 rows are displayed in the table. 
The first six rows containig 6 gene identifiers were obtained with 3 columns. 
```{r}
# Read the downloaded file
gene_data <- read.table("geneexpression.tsv",     # path to the file 
                        header=TRUE,       # indicates first row of the file contains column names 
                        sep = "\t",        # \t is the standard for tsv files 
                        row.names = 1,    #indicates that first colum contains row names
                        stringsAsFactors = FALSE )  # keep as plain character strings

# display the first 6 genes of the file
head(gene_data, 6) # 6 inidcates number of rows to be displayed 
```
Step 2: Generating a new column with mean value of other columns and displaying 6 genes 

A new column called meanexpression, which contain the mean value of other columns was created in the table using rowMeans function which calculate the mean of other columns for a given row. The output was saved in meanexpression cloumn under gene_data object. To display the output, the rows 1:6 values were selected to display first six genes and column 1 and ncol(last column) was selected to display the first and the last column (meanexpression column). 
The column for mean value of other columns were successfully generated. 

```{r}
gene_data$meanexpression <- rowMeans(gene_data) # rowMeans calculate means across all columns for each row
gene_data[1:6, c(1, ncol(gene_data))] # 1:6 selects forst 6 genes of the data 
                                      # c(1, ncol(gene_data) selects the first and the last column 
```
Step 3:  Listing the 10 genes with the highest mean expression

Firstly, the meanexpression in gene_data were ordered in the descending order using order function followed by negative (-) symbol just before the gene_data$meanexpression and then it was saved in gene_data_sorted file. After that, first 10 rows containing 10 genes with highest meanexpression were displayed in gene_data_sorted using a drop argument to ensure the datas are obtained in table format and not in vector format. 
The 10 genes with the highest mean expression was generated. The gene with the highest mean expression value was ENSG00000198804.2_MT-CO1	with mean value of 529317.3. 

```{r}
# order the "meanexpression" of the gene-data in descending order and save in gene_data-sorted file
gene_data_sorted <- gene_data[order(-gene_data$meanexpression), ] # order(-gene_data$meanofcolumns) sorts the meanexpression in descending order
# show 10 genes with the highest mea expression values 
gene_data_sorted[1:10, "meanexpression", drop = FALSE] # drop =FALSE ensures datas are expressed in data frame format and not vector 
```
Step 4. Determining the number of genes with a mean <10

to determine the genes with the meanexpression less than 10, firstly the logical vector was created that tests each gene's mean expression value where the line checks for every gene in the dataset whethere its meanexpression value is less than 10. So, if the value is less than 10, the result is true and if it is not, the result is FALSE. The output was then saved in gene_data_mean_10. Then to check how many genes meet the condition of meanexpression < 10, the sum() function was used. As the TRUE is treated as 1 and FALSE as 0 in R, summing the logical vector gives the total number of genes with meanexpression value less than 10. 
A total of 35,988 genes contain the mean expression value less then 10. 

```{r}
# create logical vectors for genes with meanexpression <10
gene_data_mean_10 <- gene_data_sorted$meanexpression <10 
# count the total number of genes with mean <10 
sum(gene_data_mean_10) # sum was used for summing all the logocal vectors
```
Step 5: Making a histogram plot of the mean values 

A histogram was generated using hist() function to represent the distribution of the mean expression value of the genes. The hist() function takes the meanexpression column from the gene_data dataset and plot its frequency distribution, where the gene_data$expression specifies the data to plot, xlab="mean expression", add descriptive label to the x-axis and main="mean value of gene expression" add the title for the plot. 
As shown in the graph, the majority of the genes have a very low mean expression as indicated by the tallest bin near zero. However, there are a few genes which has extremely higher mean expression (5e+05) but appeared to be empty due to the dominance by the tall bar for low gene expression. 
```{r}
# Step 6: Histogram of mean values of gene expression
data(gene_data)
hist(gene_data$meanexpression, # data to be plotted
     xlab="Mean expression", # label for the x-axis 
     main="Mean value of gene expression") # Main title of the histogram 
```
GROWTH DATA INTERPRETATION 

Step 6: Importing the csv file into an R object

The csv raw file for growth data was downloaded using download.file function and was saved locally as growth_data. then the file was imported into R as a data frame using the read.csv() function. This command reads the CSV file and then the file was stored in the growth_data object. Following that the colnames() function was used to display all the column names of the data set. This allowed the identification of all the column headers present in the imported dataset. 
The dataset contain six columns namely site, TreeID, Circumf_2005_cm, Circumf_2015_cm, and Circumf_2020_cm. 

```{r}
# download the growth data file

download.file("https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/growth_data.csv", destfile =  "growthdata.csv")

# Read file 
growth_data <- read.csv("growthdata.csv")
colnames(growth_data)
```
Step 7: Calculating the mean and standard deviation of tree circumference at the start and end of the study at both sites.

Firstly, the mean and standard deviation for each combination of site and year was calculated to summarize the tree circumference at the start (2005) and end (2020) of study for both sites. For this the data was subset by using logocal codition northeast and southwest site followed by application of mean() fuction to calculate the average circumference and the sd() function for calculating the mean deviation. Similarly, the steps were repeated for the northeast end (2020), southwest start(2005), southwest end (2020). After all the calculations were completed, they were combined into one clear summary atble using the data.frame() function and finally the table was displayed.This provided the summary of the average tree size and their variabilitiy at each site over the period of study. 
As shown in the table, the mean of the tree circumference increased at both the sites from 2005 to 2020. For the northeast site, the mean increased from 5.292 cm to 54.248 cm whereas for the southwest site, the mean increased from 4.862 to 45.596 cm. 
When compared at both sites, the northeast site recorded slightly higher mean circumfernece than the southwest at both start and end of study periods which could be attributed to slightly better growth conditions or initial size advantage. 
The standard variation increased dramatically from 2005 to 2020 from 0.91 to 25.22 for northeast site and from 1.14 to 17.87 for southwest site, indicating tree sizes become more variable at the end where some trees grew larger than others at both the sites. 

```{r}
# 1. Mean for Northeast site at the start (Circumf_2005_cm)

meannortheast_start<- mean(growth_data$Circumf_2005_cm[growth_data$Site== "northeast"]) # growth_data$Site== "northeast" is logical condition for noetheast site

# 2. standard deviation for the Northeast site at the start (Circumf_2005_cm)

sdnortheast_start <- sd(growth_data$Circumf_2005_cm[growth_data$Site== "northeast"])

# 3. Mean for Northeast site at the start (Circumf_2020_cm)

meannortheast_end <- mean(growth_data$Circumf_2020_cm[growth_data$Site== "northeast"])

# 4. standard deviation for the Northeast site at the start (Circumf_2005_cm)

sdnortheast_end <- sd(growth_data$Circumf_2020_cm[growth_data$Site== "northeast"])

# 5. Mean for Southwest site at the start (Circumf_2005_cm)

meansouthwest_start<- mean(growth_data$Circumf_2005_cm[growth_data$Site== "southwest"]) # growth_data$Site== "northeast" is logical condition for southwest site

# 6. standard deviation for the Southwest site at the start (Circumf_2005_cm)

sdsouthwest_start <- sd(growth_data$Circumf_2005_cm[growth_data$Site== "southwest"])


# 7. Mean for Southwest site at the end (Circumf_2020_cm)

meansouthwest_end<- mean(growth_data$Circumf_2020_cm[growth_data$Site== "southwest"])

# 8. standard deviation for the Southwest site at the end (Circumf_2020_cm)

sdsouthwest_end <- sd(growth_data$Circumf_2020_cm[growth_data$Site== "southwest"])
 
# creating summary table for mean and standard deviation of two sites at the start and end of study period

summary_table <- data.frame(
  Site = c("Northeast", "Northeast", "Southwest", "Southwest"),
  Year = c("2005 (Start)", "2020 (End)", "2005 (Start)", "2020 (End)"),
  Mean = c(meannortheast_start, meannortheast_end, meansouthwest_start, meansouthwest_end),
  SD = c(sdnortheast_start, sdnortheast_end, sdsouthwest_start, sdsouthwest_end)
)

# Displaying the table
summary_table
```
step 8: Making a box plot  for the circumference at the start and end of the study at both sites.

Boxplots was created to represent the distribution of the tree circumferences at the start and end of the study for both the sites. Firstly, the datas were splitted into Northeast and Southwest. Then, the boxplot() function was used to display the tree circumferences for both sites at both time points. To plot the box for both the sites, multiple datatsets were placed one after another inside the same boxplot() function,  separated by commas. This automatically placed the boxplots side by side in the same figure for easy comparison.

For both northeast and southwest, the tree circumference increased substantially from 2005 to 2020, indicating significant tree growth over period of time. The northeast exhibited slightly higher median and tree circumference than southwest site at both the time points, indicating that trees in the northeast experience slightly greater growth. The data was spreaded largely in 2020 box as shown by interquartile range, reflecting greater variation in tree sizes after 15 years of growth. This indicated that some trees grew significantly larger than others which could be attributed to environmental factors. Overall, the box displayes increased tree size and variability over time at both study sites. 

```{r}
# Splitting of data into Northeast and Southwest

northeast <- growth_data[growth_data$Site == "northeast", ]
southwest <- growth_data[growth_data$Site == "southwest", ]


# Creating boxplots for two different sites at the start and end of the study periods

boxplot(northeast$Circumf_2005_cm, northeast$Circumf_2020_cm,
        southwest$Circumf_2005_cm, southwest$Circumf_2020_cm,
        names = c("NE Start","NE End", "SW Start", "SW End"), # Provides the names to each boxplot
        col = c("red", "blue", "red", "blue"), # provides red color to the boxplot at the start and blue color to the boxplot for the end for both the study sites
        ylab = "Tree Circumference (cm)",
        main = "Tree Circumference at Start and End by Site")
        
```
Step 9: Calculating the mean growth over the last 10 years at each site.

To determine the mean tree growth over the last 10 years(from 2010 to 2020) at each study site, the difference in circumference between 2020 and 2010 was calculated for each tree and the output was saved in the new column called growth_10_years under object growth_data. Then, data were sepa
rated by site to calculate the mean growth for each location. Following that, the mean () function was used separately to the northeast and southwest subsets to obtain the average tree growth over the 10-year period for each site.

```{r}
# calculate growth data from 2010 to 2020

growth_data$ growth_10_years <- (growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm)

# Extracting 10-year growth values for northsite
North_east_growth_data <- (growth_data$growth_10_years[growth_data$Site == "northeast"])

# calculating mean for 10-year growth values of northwest
North_east_mean_growth_data <- mean(North_east_growth_data)
North_east_mean_growth_data

# Extracting 10-year growth values for southwest
South_west_growth_data <- (growth_data$growth_10_years[growth_data$Site == "southwest"])

# calculating mean for 10-year growth values of southwest

Southwest_mean_growth_data <- mean (South_west_growth_data)
Southwest_mean_growth_data
```

Step 10: Using the t.test to estimate the p-value that the 10 year growth is different at the two sites

The p-value for the two sites were determined using t-test function. The north_east_growth_data which contains the last 10 years growth data for northeast site and South_west_growth_data which contains last 10 years growt data for southwest site were used to compare the growth between the two sites.The study observed higher mean 10-year growth of 42.94 at the northeast site compared to soutwest site (35.49 cm) but the growth difference was not statistically significant at 5% significance level (t-value = 1.8882, df = 87.978, and p-value = 0.06229). 
```{r}
t.test(North_east_growth_data,South_west_growth_data)
```
 
PART 2: Examining the biological sequence diversity 

step 1: Downloading the Saprospirale and ecoli sequences, counting the coding sequences (CDS) and comparing the sequences between the two organisms. 

The codes such as sequinr and R.utils were used for smooth downloading, unzipping and reading of CDS data. 
The compressed FASTA files (.fa.gz) for both the E.coli and Saprospirales were downloaded from the ENSEMBL website using the download.file() function.
Since the downloaded file was in gzip format, the gunzip() function was used from the R.utils package to extract the .fa files. The overwrite argument was used to overwrite the unzipped file as R usually refuses to overwrite it. The read.fasta() function from the sequinr package was used to load the sequence into R as lists, where each element represents a coding sequence. Then, the length() function was applied on each list of sequences to determine the total number of CDS for each organism. Finally the table was created using data.frame() function with two columns , one for the organism name and one for the number of coding sequences, and displayed it as a table. 
 
As shown in the table,  the Saprospirales contain 4527 coding sequences whereas the E.coli contains 4239 coding sequences. Therefore, the saprospirales has slightly more number of coding sequences than E.coli, which indicates that saprospirales have comparatively more complex genome which could be attributed to their adaptation to diverse environmental conditions. 

```{r}
suppressPackageStartupMessages({
  library("seqinr") # is a package designed to process and analyse sequence data.
  library("R.utils") # general utilities like zip and unzip
})
# loading e.coli and Saprospirales data 
library("R.utils")

# Download FASTA file for saprospirales 
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_58_collection/saprospirales_bacterium_gca_003448025/cds/Saprospirales_bacterium_gca_003448025.ASM344802v1.cds.all.fa.gz"
download.file(URL,destfile="saprospirales_cds.fa.gz")

# Download FASTA file for E.coli

URL="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")

# Uncompress the FASTA files for e.coli

gunzip("ecoli_cds.fa.gz", overwrite  = TRUE) 
# overwrite = TRUE tells R to replace the uncompressed .fa file if it already exist in the working directory

# Uncompress the FASTA file for Saprospirales 
gunzip("saprospirales_cds.fa.gz", overwrite  = TRUE) # gunzip is used to uncompress the file

# list files 

list.files()

# Read the FASTA sequences for e.coli 
library("seqinr")
e.coli_cds <- seqinr::read.fasta("ecoli_cds.fa") # read.fasta is used to read the FASTA file 

# Read the FASTA sequences for Saprospirales
saprospirales_cds <- seqinr::read.fasta("saprospirales_cds.fa")

# Count the number of coding sequences for e.coli 
e.coli_number <- length(e.coli_cds)

# Count the number of coding sequences for Saprospirales
saprospirales_num <- length(saprospirales_cds)

# create table 
cds_table <- data.frame(
  Bacteria = c("E.coli", "Saprospirales"),
  CDS_count = c(e.coli_number, saprospirales_num)
)
cds_table

```
Step 2: Determing and comparing the total coding DNA between the two organisms.  

To determine the total length of the the sequences in both organisms, the length of each coding sequences was extracted from the respective organisms' cds using summary() function and the first column of the summary was converted to the numeric value using as.numeric() function. Then the total length of all the genes were calculated by summing these values using the sum() function. Then, to display the compiled result of two organisms, the table was created using data.frame() function with columns for organism, number of CDA and total coding DNA.  

It was observed that the Saprospirales has greater total coding DNA (4200321) than that of E.coli (3978528). This indicates that Saprospirales has a larger or more complex genome with potentially more genes or longer coding sequences, which could be attributed to adaptation to a diverse environmental condition. 

```{r}

# Calculate the CDS length for e.coli 
e.coli_cds_length<- as.numeric(summary(e.coli_cds)[,1]) # 1 select the first column of the matrix 

# Calculate the CDS length for Saprospirales 
saprospirales_cds_length <- as.numeric(summary(saprospirales_cds)[,1])

# sum total coding DNA of E.coli
e.coli_total_cds_length <- sum(e.coli_cds_length)

#Sum total coding DNA of Saprospirales 
saprospirales_total_cds_length <- sum(saprospirales_cds_length)

# create table for the total coding DNA for both organisms
cds_table <- data.frame(
  Bacteria = c("E.coli", "Saprospirales"),
  CDS_count = c(e.coli_number, saprospirales_num),
  total_coding_DNA = c(e.coli_total_cds_length, saprospirales_total_cds_length)
)
cds_table

```

Step 3: Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms. What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms.

The length of each coding sequences was extracted from the respective organisms' cds using summary() function and the first column of the summary was converted to the numeric value using as.numeric() function. Then the boxplot() was used to display the distribution of coding aequence lengths in both organisms where red color represented the E.coli coding seuqnece length and blue color represented the Saprospirales coding sequence length.Following that the mean() and median() function was used to summarize the central tendency of CDS lengths. To compare the result between two organisms, the table was generated using data.frame() function, representing the value of CDS count, total coding DNA, mean and median CDS length.
The mean coding sequence length of E.coli is 938.5534 and the median is 831. Whereas in case of Saprospirales, the mean coding sequence length is 927.8376 and the median is 690. This shows that mean CDS lnegth of E.coli is slightly higher than Saprospirales, indicating that on average E.coli genes are bit longer. Similarly, the median CDS length is substantially higher than Saprospirales, indicating the most E.coli genes are moderately long, whereas Saprospirales has a larger proportion of shorter genes. 

```{r}

# Calculate the CDS length for E.coli 
e.coli_cds_length<- as.numeric(summary(e.coli_cds)[,1]) # 1 select the first column of the matrix 

# Calculate the CDS length for Saprospirales 
saprospirales_cds_length <- as.numeric(summary(saprospirales_cds)[,1])


# Generate box plot for E.coli and Saprospirales
boxplot(list(E.coli = e.coli_cds_length, Saprospirales = saprospirales_cds_length),
        col = c("red", "blue"),
        ylab = "CDS Length (bp)",
        main = "Distribution of Coding Sequence Lengths")

# Calculate mean and median of E.coli 
e.coli_mean_cds_length <- mean(e.coli_cds_length)
e.coli_median_cds_length <- median(e.coli_cds_length)

# Calculate mean and median of Saprospirales
saprospirales_mean_cds_length <- mean(saprospirales_cds_length)
saprospirales_median_cds_length <- median(saprospirales_cds_length)

# Create a table 
cds_table <- data.frame(
  Bacteria = c("E.coli", "Saprospirales"),
  CDS_count = c(e.coli_number, saprospirales_num),
  total_coding_DNA = c(e.coli_total_cds_length, saprospirales_total_cds_length),
  mean_cds_length = c(e.coli_mean_cds_length, saprospirales_mean_cds_length),
  median_cds_length = c(e.coli_median_cds_length, saprospirales_median_cds_length)
)
cds_table


```
Step 4: Calculating the frequency of DNA bases and aminoacids in the total coding sequences for both organisms and generating barplots.

The unlist() function was used to turn the list of CDS into a single long vectors of single characters for both the organisms. Then the frequency of dna bases in total coding sequences was calculated using count() function and to obtain single nucleotide count, 1 was used. To obtain the bar plot, the table was plotted using data.frame for nucleotide frequency for both organisms followed by visualisation using the barplot with xlab="nucleotides", ylab="Count", and main="Nucleotide Frequency in CDS". The x-axis represented the nucleotides (A, C, G, T), the y-axis showed their frequencies, and the plot title indicated that it displays the nucleotide composition of organism's coding sequences.

The lapply which applies the translate function to each element of the cds list was used to transalte the dna to protein, resulting in specific organism's prot, a list iif translated protein seuences. Then, unlist() was used ti give a long aminoacid vector. Following that, the aminoacids frequency were counted using the count() function and table was generated for aa frequency of both organisms. Then, the barplot was generated for the aminoacid frequency. 

The higher frequency of adenine (A) was recorded higher in Saprospirales while E.coli recorded slightly higher frequency of guanine(G) content. Cytosine (C) and Thymine (T) levels were comparable between the two species. 

In both the organisms, Leucine (L) and alanine (A) were the most abundant amino acids , followed by glycine (G), serine (S), and valine (V). While overall profiles were similar, E.coli exhibited higher frequencies of acid residues such as aspartic acid (D) and glutamic acid (E), along with slightly more histidine (H). In contrast, Saprospirlaes showed relatively higher frquencies of valine (V), glutamine (Q), and tryptophan(W)

```{r}
# Unlist the E.coli cds 
e.coli_dna <- unlist(e.coli_cds)

# Calculate frequency of dna bases in total coding sequences for e.coli
e.coli_dna_freq <- count(e.coli_dna, 1)   #1 =word size (single nucleotide count)

# Unlist the Saprosipirales cds sequences 
saprospirales_dna <- unlist(saprospirales_cds)

# Calculate the frequency of dna bases in total coding sequences for Saprospirales
saprospirales_dna_freq <- count(saprospirales_dna, 1)   # 1 = word size (single nucleotide count)

# Combine into a data frame for plotting
Dna_freq_df <- data.frame(
  Base = c("A","C","G","T"),
  E.coli = e.coli_dna_freq,
  Saprospirales = saprospirales_dna_freq
)
Dna_freq_df

# Generate bar plot for nucleotide frequency 
barplot(
  height = rbind(as.numeric(Dna_freq_df$E.coli.Freq), as.numeric(Dna_freq_df$Saprospirales.Freq)),
  beside = TRUE,      # Place the bar for E.coli and Saprospirales side-by-side
  names.arg = base_freq_df$Base, # Tells R to represent nucleotide bases (AGTC) at the x-axis 
  col = c("red", "blue"), # Tells R to give red color to E.coli and blue to saprospirales
  main = "Nucleotide Frequency in CDS", 
  ylab = "Count",
  xlab="nucleotides"
)
legend("topright", legend = c("E. coli", "Saprospirales"), fill = c("red", "blue"))
# topright specify the position of the legend, legend =c("E. coli", "Saprospirales") are the labels of the legend and fill = c("red", "blue") fits the color of the boxes 

# Translate the sequences for e.coli
e.coli_prot <-lapply(e.coli_cds, translate) # lapply applies the translate function to each element of the cds list, resulting in specific organism_prot, a list of translated protein sequences

# Unlist aminoacids of E.coli
e.coli_prot_unlist <- unlist(e.coli_prot)

# Count aminoacids 
aa_alphabet <- c("A","R","N","D","C","Q","E","G","H","I","L","K","M","F","P","S","T","W","Y","V")

# Calculate E.coli aminoacid frequency
e.coli_aa_freq <- count(e.coli_prot_unlist, wordsize=1, alphabet=aa_alphabet)
e.coli_aa_freq


# Translate the sequences for Saprospirales
saprospirales_prot <- lapply(saprospirales_cds, translate)
 
# Unlist aminoacids of Saprospirales
saprospirales_prot_unlist <- unlist(saprospirales_prot)

# Calculate the frequency of aa of Saprospirales 
Saprospirales_aa_freq <- count(saprospirales_prot_unlist, wordsize=1, alphabet=aa_alphabet)
Saprospirales_aa_freq


# combine the aa frequency into data.frame for plotting 
aa_freq_df <- data.frame(
  AA = aa_alphabet,
  E_coli = e.coli_aa_freq,
  Saprospirales = Saprospirales_aa_freq
)
aa_freq_df

# Generate bar plot for nucleotide frequency 
barplot(
  height = rbind(as.numeric(aa_freq_df$E_coli.Freq), as.numeric(aa_freq_df$Saprospirales.Freq)),
  beside = TRUE,
  names.arg = aa_freq_df$AA, # Tells R to represent aa texts at the x-axis 
  col = c("red", "blue"),
  main = "Nucleotide Frequency in CDS",
  ylab = "Count", 
  xlab = "Aminoacids"
)
legend("topright", legend = c("E. coli", "Saprospirales"), fill = c("red", "blue"))
```
Create a codon usage table and quantify the codon usage bias among all coding sequences. Describe any differences between the two organisms with respect to their codon usage bias. Provide charts to support your observations.

The uco() function from the sequinr package was applied to count all codons (3-base sequences) in the DNA sequences. The codons were sorted using order() function to ensure that the codon table is consistent for both organisms. Following that, the table was created to compare the codon counts between the organisms, where each row is a codon with counts in E.coli and Saprospirales. 
Then, the index="rscu" was employed to compute relative synonymous codon usage which is a measure of codon usage bias. Here, the RSCU > 1 indicates that codon is used more frequently than expected for the amino acid while RSCU <1 indicates that codon is used less frequently than expected. The data.frame was kept TRUE to convert the result into plotting. 

The barchart was generated for codon usage bias  using rbind() function which makes matrix with rows = organisms and columns = codons. Then beside=TRUE was used to keep the charts of two organisms side by side for comparison and legend was added to the chart. 

As shown in the chart, the E.coli shows stronger peaks for certain codons, indicating higher codon usage bias while Saprospirales exhibit more even distributions across synonymous codons, suggesting weaker codon preference. Moreover, some of the codons which are less used in saprospirales and simlarly, the codons which are are more used in Saprospirales are comparatively less used in E.coli. This shows that the two organisms are adapeted to different tRNA pools and translational pressures. 

```{r}
# Determining codon usage for e.coli
e.coli_codon_usage <- uco(e.coli_dna)
e.coli_codon_usage <- e.coli_codon_usage[order(names(e.coli_codon_usage))]  # sort codons
e.coli_codon_usage

# Determining codon usage for saprospirales
saprospirales_codon_usage <- uco(saprospirales_dna) # uco() returns counts of each codon.
saprospirales_codon_usage <- saprospirales_codon_usage[order(names(saprospirales_codon_usage))]  # sort codons
saprospirales_codon_usage

# Creating table for codon_usage for E.coli and Saprospirales
codons <- names(e.coli_codon_usage)
bacteria_codon_table <- data.frame(
  Codon = codons,
  E.coli_count = as.numeric(e.coli_codon_usage),
  Saprospirales_count = as.numeric(saprospirales_codon_usage)
  )
bacteria_codon_table

# Calculation of RSCU values for E.coli 
e.coli_codon_usage_bias <- uco(e.coli_dna, index="rscu", as.data.frame=TRUE)

# Calculation of RSCU values for Saprospirales
saprospirales_codon_usage_bias <- uco(saprospirales_dna, index="rscu", as.data.frame=TRUE)
e.coli_codon_usage_bias


# generate barchart for codon usage bias for e.coli and saprospirales
RCSU_matrix <-rbind(E.coli = e.coli_codon_usage_bias$freq, Saprospirales = saprospirales_codon_usage_bias$freq)
barplot(RCSU_matrix,
        beside = TRUE,
        names.arg = e.coli_codon_usage_bias$AA,
        legend.text = TRUE,
        ylab = "codon usage bias (RSCU)",
        xlab = "Codons",
        main = "Codon usage frequency usage comparison"
        )
grid()

```
Step 6: In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most over- and under-represented k-mers in your organism of interest. Are these k-mers also over- and under-represented in E. coli to a similar extent? Provide plots to support your observations. Why do you think these sequences are present at different levels in the genomes of these organisms?
```{r}
# Translate the sequences for e.coli
e.coli_prot <-lapply(e.coli_cds, translate) # lapply applies the translate function to each element of the cds list, resulting in specific organism_prot, a list of translated protein sequences

# Unlist aminoacids of E.coli
e.coli_prot_unlist <- unlist(e.coli_prot)

# Calculate 3-mer (amino acid triplet) frequency in E. coli protein sequences
e.coli_prot_3_count <- count(e.coli_prot_unlist, wordsize = 3, alphabet = aa_alphabet)

e.coli_prot_4_count <- count(e.coli_prot_unlist, wordsize = 4, alphabet = aa_alphabet)

w.coli_prot_5_count <-  count(e.coli_prot_unlist, wordsize = 5, alphabet = aa_alphabet)

# Structure the e.coli_prot_count 
str(e.coli_prot_count)

# Translate the sequences for Saprospirales
saprospirales_prot <- lapply(saprospirales_cds, translate)
 
# Unlist aminoacids of Saprospirales
saprospirales_prot_unlist <- unlist(saprospirales_prot)

# Calculate 3-mer (amino acid triplet) frequency in Saprospirales protein sequences
Saprospirales_prot_3_count <- count(saprospirales_prot_unlist, wordsize = 3, alphabet = aa_alphabet)
Saprospirales_prot_4_count <- count(saprospirales_prot_unlist, wordsize = 4, alphabet = aa_alphabet)
Saprospirales_prot_5_count <- count(saprospirales_prot_unlist, wordsize = 5, alphabet = aa_alphabet)

# Function to get top over- and under-represented k-mers
top_kmers <- function(kmer_counts, N = 10){
  kmer_sorted <- sort(kmer_counts)
  under <- head(kmer_sorted[kmer_sorted > 0], N)  # smallest non-zero counts
  over  <- tail(kmer_sorted, N)                   # largest counts
  list(over = over, under = under)
}

# Saprospirales 3-mers
Saprospirales_k3_top <- top_kmers(Saprospirales_prot_3_count)
Saprospirales_k3_top

# E. coli 3-mers
k3_top_ecoli <- top_kmers(e.coli_prot_3_count)

# Get counts of Saprospirales over- and under-represented 3-mers in E. coli
compare_kmers <- function(target_kmers, organism_kmer_counts){
  organism_kmer_counts[names(target_kmers)]
}

# Over-represented in Saprospirales
k3_over_in_ecoli <- compare_kmers(k3_top_sapro$over, e.coli_prot_3_count)

# Under-represented in Saprospirales
k3_under_in_ecoli <- compare_kmers(k3_top_sapro$under, e.coli_prot_3_count)

over_df <- data.frame(
  Kmer = names(k3_top_sapro$over),
  Sapro = as.numeric(k3_top_sapro$over),
  E_coli = as.numeric(k3_over_in_ecoli)
)

barplot(
  t(as.matrix(over_df[,2:3])),
  beside = TRUE,
  col = c("blue", "red"),
  names.arg = over_df$Kmer,
  las = 2,  # rotate x-axis labels
  ylab = "Count",
  main = "Top 10 Over-represented 3-mers: Saprospirales vs E. coli"
)
legend("topright", legend=c("Saprospirales", "E. coli"), fill=c("blue","red"))



# Structure the e.coli_prot_count 
str(Saprospirales_prot_count)


```

